### PR TITLE
Api 27078 526 v2 allow MM-YYYY date format

### PR DESCRIPTION
**Pull request by rockwellwindsor-va**
Handle submitting approximate date values with only month and year, so that the Veteran can provide an approximate date where needed.

## Summary
* Updates JSON files to allow for date string pattern of MM-YYYY or MM-DD-YYYY for the following values: disabilities.approximateDate, secondaryDisability.approximateDate, confinements start and end dates and separationSeverancePay.datePaymentReceved
* Updates JSOn files to allow for date string pattern of MM-YYYY in toxic exposure values and treatment.startDate
* Updates validation file to handle dates of those formats when comparing dates.  Also updates logic so when comparing approximate dates allowing for them to be equal.
* Updates swagger file to reflect new REGEX patterns and examples for those dates
* Updates RSpec tests where needed to handle new date formats or new date format comparison scenarios.

## Related issue(s)
[API-27078](https://jira.devops.va.gov/browse/API-27078)

## Testing done
* RSpec

## What areas of the site does it impact?
modified:   modules/claims_api/app/controllers/claims_api/v2/veterans/disability_compensation_controller.rb
modified:   modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb
modified:   modules/claims_api/app/swagger/claims_api/v2/dev/swagger.json
modified:   modules/claims_api/config/schemas/v2/526.json
modified:   modules/claims_api/config/schemas/v2/request_bodies/disability_compensation/example.json
modified:   modules/claims_api/config/schemas/v2/request_bodies/disability_compensation/request.json
modified:   modules/claims_api/lib/claims_api/v2/disability_compensation_pdf_mapper.rb
modified:   modules/claims_api/spec/fixtures/v2/veterans/disability_compensation/form_526_json_api.json
modified:   modules/claims_api/spec/lib/claims_api/v2/disability_compensation_pdf_mapper_spec.rb
modified:   modules/claims_api/spec/requests/v2/veterans/disability_compensation_request_spec.rb
modified:   spec/support/schemas/claims_api/v2/forms/disability/submission.json

## Acceptance criteria
- [X]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [X]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [X]  Documentation has been updated (link to documentation) (RSWAG docs in dev)
- [X]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 1]
```diff
@@ -244,6 +245,7 @@ def validate_form_526_disability_secondary_disability_classification_code_matche
         )
       end
 
+      # TO-DO:Fix the dates here to handle the new format
```
Comment (ryan-mcneil) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:248]: just confirming this TODO is intended for later

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 2]
```diff
@@ -576,6 +579,60 @@ def raise_exception_if_value_present(val)
...
+        # figure out which one has the day and remove it
+        if date_has_day?(begin_date)
+          begin_date = remove_chars(begin_date.dup)
+        elsif date_has_day?(end_date)
+          end_date = remove_chars(end_date.dup)
+        end
+        Date.strptime(begin_date, '%m-%Y') > Date.strptime(end_date, '%m-%Y') # = is ok
+      end
+
+      def date_is_valid_against_now_after_check_on_format?(date)
```
Comment (ryan-mcneil) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:608]: Not necessary for my approval, but consider rethinking some of these method names. It's a little difficult to determine what they're meant to do (from the outside looking in, at least)

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb [hunk 3]
```diff
@@ -576,6 +579,60 @@ def raise_exception_if_value_present(val)
...
+      end
+
+      # just need to know if day is present or not
+      def date_has_day?(date)
+        case date.length
+        when 7
+          false
+        when 10
+          true
+        end
```
Comment (ryan-mcneil) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:626]: Consider shortening this method using a ternary.

`date.length === 10` would do the same thing as the entire body of the method, unless you're expecting that `nil` value for the case where it's neither 7 or 10.

**Review by ryan-mcneil — APPROVED** [review]
Approved, but let me know if you implement any of my recs and you need me to reapprove!

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb (hunk 3 above)
Comment (rockwellwindsor-va) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb:626]: I initially tried a ternary but  mine was searching for 10 or 7 specifically which it is not needed so this is great, thanks for pointing this out! making this change now.

File: Gemfile.lock [hunk 4]
```diff
@@ -1203,4 +1203,4 @@ RUBY VERSION
    ruby 3.2.2p53
 
 BUNDLED WITH
-   2.4.10
+   2.4.10
```
Comment (mchristiansonVA) [inline Gemfile.lock]: Gemfile.lock should be removed from this branch.

Comment (rockwellwindsor-va) [reply Gemfile.lock]: Ya working on it now, thanks for the heads up!

File: modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb (hunk 1 above)
Comment (rockwellwindsor-va) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/disability_compensation_validation.rb]: Thanks for the heads up!

File: Gemfile.lock (hunk 4 above)
Comment (rockwellwindsor-va) [reply Gemfile.lock:1206]: ok after three tries this should be cleaned up

Comment (mchristiansonVA) [reply Gemfile.lock:1206]: Still seeing this file in your branch/github list of changed files.

**Review by mchristiansonVA — COMMENTED** [review]
Wondering if we should standardize all the date fields at this point? Currently we've got a mix of MM-YYYY, MM-DD-YYYY, and the previous format YYYY-MM-DD. Is there any reason not to standardize on MM-YYYY & MM-DD-YYYY for all date fields?

If you take a look at "serviceInformation" in 526.json, the "servicePeriods" section has "activeDutyBeginDate" & "activeDutyEndDate" in YYYY-MM-DD. The "confinements" section just below it uses MM-DD-YYYY for the "approximateBeginDate" & "approximateEndDate".

**Pull request closed without merging**